### ISSUE TITLE
`unsound-assignment` flags every assignment of `int ** int`

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ty#4358]
### Summary

The exponent of two `int`s is a `float | int`, but [Typeshed annotates it as `Any` to avoid false positives](https://github.com/python/typeshed/blob/21dff5c0ca9a6cc18f70dac51b36e1f08e30be5d/stdlib/builtins.pyi#L322-L325). With `unsound-assignment` activated in ty, every assignment of such an exponent becomes a false positive, which is unhelpful. It would be more helpful and more correct to treat the type as `float | int` when `unsound-assignment`  is active. [Example](https://play.ty.dev/f604f11f-b3fa-49f0-962e-c48de6189d69):
```console
$ cat >main.py <<'# EOF'
def f(x: int, y: int) -> None:
    _: float = x ** y
# EOF

$ ty check --error unsound-assignment main.py
error[unsound-assignment]: Unsound assignment
 --> main.py:2:16
  |
2 |     _: float = x ** y
  |        -----   ^^^^^^ Inferred as `Any`
  |        |
  |        Expected a subtype of `float` because of this annotation
info: `Any` is assignable to `float`, but not a subtype of `float`
help: Consider using an `assert` to narrow the type before assigning it

Found 1 diagnostic
```

### Version

ty 0.0.74 (00199f0aa 2026-08-22)

**Comment by AlexWaygood** [comment]
Thanks for the report! I agree that this is suboptimal, but I'm not sure there's really anything to be done here. I don't think it's plausible to override typeshed's signature for `int.__pow__` unconditionally so that it returns `float` -- there's too much code out there that knows that it only ever has positive `int`s, and so knows that the result of `x ** y` will always be an `int`, and so would have false-positive errors if we did that. And I also don't think we should selectively override typeshed's signature for `int.__pow__` based on whether or not a certain rule is enabled. That feels like it could lead to very confusing behaviour. Overriding one of typeshed's signatures is anyway something that we should do very rarely, since it means that we're less compatible with other type checkers, and it doesn't scale.

The reason why this rule is disabled by default is that it's targeted towards users who want to have additional soundness checks from their type checker that go above and beyond the standard typing system rules. The tradeoff there is that you have to accept a certain amount of busywork will be required from you to fix the diagnostics from this rule, since typeshed's signatures etc. are designed assuming teh normal typing system rules. So here, I think I would simply recommend following the `help` suggestion and using an `assert` to narrow the type. Or, if the rule is too noisy for you, just don't enable it.

**Comment by dscorbett** [comment]
How can I use `assert` to narrow the type without first assigning it?

**Comment by AlexWaygood** [comment]
```py
tmp = x ** y
assert isinstance(tmp, float)
_: float = tmp
```

**Comment by carljm** [comment]
Ideally typeshed would annotate this return type as something like `(float | int) & Any`, which effectively expresses "either float or int, with optimistic gradual assignability". That would give the same lax behavior for non-strict users, and avoid the false positive `unsound-assignment` when assigning to `float`. But that requires intersection types to be standardized.

Short of that, I don't see what ty can do here; the `unsound-assignment` feature is working as advertised. I am actually tempted to do the conditional typeshed patch, but I agree that doesn't really scale -- there are a lot of pragmatic `Any` in typeshed, which just means `unsound-assignment` is inherently a bit painful to use.

Closing as "not planned" for now since it's not actionable for ty, but this is not meant to preclude further discussion here.
